[WFCORE-5464] Upgrade DMR to be able to resolve expressions from env …

kabir opened the pull request:
…vars

https://issues.redhat.com/browse/WFCORE-5464

jmesnil commented on management/WFCORE-5464_Check_Env_Vars_On_Expression_Resolution.adoc:
```diff
@@ -0,0 +1,101 @@
+
+This means that if the expression can be resolved by both a system property and an environment variable, the system property takes precedence, as per the `Default ConfigSources` section of the MicroProfile Config specification. The other ConfigSources mentioned are not relevant for this use case.
+
+When converting between the expression name and the environment variable we will only use option 3 of the `Environment Variables Mapping Rules` from the MicroProfile Config specification. The aim is to avoid having to do too much string manipulation and too many lookups when resolving expressions, and to my knowledge environment variables available on OpenShift (which is the main target for this  ...
```
this naming convention is usually called SCREAMING_SNAKE_CASE 😆

vsalbaba commented:
I think there is some danger in reading an existing environment variable by mistake after upgrading.

If I have an environment variable `COMMON_VAR_NAME=foo` already in use
and I have `${common-var-name:bar}` in wildfly configuration,
and I expect the value to resolve to default,
when I upgrade to new wildfly version
then the value of `${common-var-name:bar}` will be sudenly `foo`

I think we should at least add warning about possible conflicts into the update notes or docs

kabir commented:
@vsalbaba Good catch, I have updated the documentation PR https://github.com/wildfly/wildfly/pull/14376, and added a "Other Documentation Concerns" to this analysis document

vsalbaba commented:
> I think there is some danger in reading an existing environment variable by mistake after upgrading.

I've had one more issue raised about this, it would be helpful to log which attributes were overridden from an environment variable.

This would be imho a nice solution to "how to debug that my attributes suddenly changed after wildfly update" problem.

kabir commented:
Regarding the log question, we've discussed a bit internally and feel that is a quite a corner case scenario. Turning on debug logging isn't really much easier than inspecting the environment, and that the documentation should suffice.

Also, the resolution happens in jboss-dmr which is a zero dependency library (its [pom](https://github.com/jbossas/jboss-dmr/blob/master/pom.xml)), and I doubt @dmlloyd would accept adding the jboss logging dependency to it.

dmlloyd commented:
I think it would probably be OK to add logging to this project via `jboss-logging`.  That's my usual one exception to having no dependencies.  Anyway it is likely that any run time environment which contains DMR would also contain `jboss-logging`.

kabir commented:
@jmesnil @dmlloyd @vsalbaba I've added the logging to https://github.com/wildfly/wildfly-proposals/pull/400 and updated https://github.com/wildfly/wildfly-core/pull/4633 to account for it
